src/control/cad.js

```diff
@@ -253,103 +261,265 @@ class CadControl extends Control {
+    const minVertLine = new LineString([
+      [xMin[0], -20037508.342789],
+      [xMin[0], 20037508.342789],
+    ]);
+    minVertLine.rotate(this.map.getView().getRotation(), xMin);
+    const maxVertLine = new LineString([
+      [xMax[0], -20037508.342789],
+      [xMax[0], 20037508.342789],
+    ]);
+    maxVertLine.rotate(this.map.getView().getRotation(), xMax);
+    const minHoriLine = new LineString([
+      [-20037508.342789, yMin[1]],
+      [20037508.342789, yMin[1]],
+    ]);
+    minHoriLine.rotate(this.map.getView().getRotation(), yMin);
+    const maxHoriLine = new LineString([
+      [-20037508.342789, yMax[1]],
+      [20037508.342789, yMax[1]],
+    ]);
+    maxHoriLine.rotate(this.map.getView().getRotation(), yMax);
+
+    // Use intersection points of the four lines to get the extent
+    const intersectTopLeft = OverlayOp.intersection(
+      parser.read(minVertLine),
+      parser.read(minHoriLine),
+    );
+    const intersectBottomLeft = OverlayOp.intersection(
+      parser.read(minVertLine),
+      parser.read(maxHoriLine),
+    );
+    const intersectTopRight = OverlayOp.intersection(
+      parser.read(maxVertLine),
+      parser.read(minHoriLine),
+    );
+    const intersectBottomRight = OverlayOp.intersection(
+      parser.read(maxVertLine),
+      parser.read(maxHoriLine),
+    );
+
+    return [
+      [intersectTopLeft.getCoordinate().x, intersectTopLeft.getCoordinate().y],
+      [
+        intersectBottomLeft.getCoordinate().x,
+        intersectBottomLeft.getCoordinate().y,
+      ],
+      [
+        intersectTopRight.getCoordinate().x,
+        intersectTopRight.getCoordinate().y,
+      ],
+      [
+        intersectBottomRight.getCoordinate().x,
+        intersectBottomRight.getCoordinate().y,
+      ],
+    ];
+  }
+
   /**
    * Draws snap lines by building the extent for
    * a pair of features.
    * @private
    * @param {Array.<ol.Feature>} features List of features.
    * @param {ol.Coordinate} coordinate Mouse pointer coordinate.
+   * @param {Array.<number>} pixel Mouse pointer pixel array.
    */
-  drawSnapLines(features, coordinate) {
+  drawSnapLines(features, coordinate, pixel) {
     let auxCoords = [];
 
+    // First get all snap points: neighbouring feature vertices and extent corners
     for (let i = 0; i < features.length; i += 1) {
       const geom = features[i].getGeometry();
       const featureCoord = geom.getCoordinates();
-
-      if (geom instanceof Point) {
-        auxCoords.push(featureCoord);
-      } else {
-        // filling snapLayer with features vertex
-        if (geom instanceof LineString) {
-          for (let j = 0; j < featureCoord.length; j += 1) {
-            auxCoords.push(featureCoord[j]);
-          }
-        } else if (geom instanceof Polygon) {
-          for (let j = 0; j < featureCoord[0].length; j += 1) {
-            auxCoords.push(featureCoord[0][j]);
+      if (featureCoord.length) {
```

Comment:
In which case this condition returns false ?

Reply:
When drawing polygons it returns a feature with an empty coordinate array, must be an OL thing

Reply:
so add a comment otherwise it will disappear some day :D